<a href="https://colab.research.google.com/github/saritmaitra/Optimizatiion_problem/blob/main/Linear_Programming.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##### **Prepared by Dr. Sarit Maitra on behalf of Dr. B. C. Roy Polytechnic(A Unit of Dr. B. C. Roy Engineering College, Durgapur)**

In [ ]:
!pip install pulp
import pulp
from pulp import *

## ***Business case & problem statement***

### ***Use case 1***

A bakery makes cakes and pies every day of a month. There is: 1 oven, 2 bakers, 1 packaging packer that works only 22 days of the month. 

- The cake requires to use the oven for 1 day and the pie requires 0.5 day. 
- Each baker needs to work for cake 0.5 days and pie 2 days. 
- Packer needs to work for cake 1 day and pie 0.5 days. 

The profit on each cake is 15 dollar and the profit on each pie is 12 dollar. 

How many should be made to maximize the profit under given conditions?

In [ ]:
# initializing model with objective function 'maximize'
model = LpProblem('Maximize_Profits', sense= LpMaximize)

# defining decision variables
C = LpVariable('C', lowBound=0, upBound=None, cat='Integer')
P = LpVariable('P', lowBound=0, upBound=None, cat='Integer')

# we define a name for the variable, values for lower and upper bound, and category type ‘Integer’. 

model += 15 * C + 12 * P # defining objective function

#defining constraints
model += 1 * C + 0.5 * P <= 30
model += 0.5 * C + 2 * P <= 60
model += 1 * C + 0.5 * P <= 22

# solving the problem
print(model.solve())

#Print model status
print('Status:', LpStatus[model.status]) # Check the status of the solution 

#Print solution variables
for variable in model.variables():
    print ('{} = {}'.format(variable.name, variable.varValue))

print('Optimal number:', value(model.objective)) # print optimized objective function


1
Status: Optimal
C = 8.0
P = 28.0
Optimal number: 456.0


**Solution: Bakery needs to produce 456 numbers of cake / month to maximize profit.**


### ***Use Case 2***
A factory is looking to hire workers, with the requirements to work 5 consecutive days and then 2 days off.

The objective is to hire the minimum number of workers and the estimated number of worker needed for each day is:
- Monday: 25,
- Tuesday: 32,
- Wednesday: 22,
- Thursday: 18,
- Friday: 24,
- Saturday: 12,
- Sunday: 14.

What would be the minimum number of workers to hire?

In [ ]:
# Mon --> X0 + X3 + X4 + X5 + X6 = 25
# Tue --> X0 + X1+ X4 + X5 + X6 = 32
# Wed --> X0 + X1 + X2 + X5 + X6 = 22
# Thu --> X0 + X1 + X2 + X3 + X6 = 18
# Fri --> X0 + X1 + X2 + X3 + X4 = 24
# Sat --> X1 + X2 + X3 + X4 + X5 = 12
# Sun --> X2 + X3 + X4 + X5 + X6 = 14

In [ ]:
#Initialize model
model = LpProblem("Minimize_Number_of_Workers", LpMinimize)

#Define decision variables
days = list(range(7))
x = LpVariable.dicts('workers_', days, lowBound=0, upBound=None, cat='Integer')

#Define model
model += lpSum([x[i] for i in days])

# Define constraints
model += x[0] + x[3] + x[4] + x[5] + x[6] >= 25
model += x[0] + x[1] + x[4] + x[5] + x[6] >= 32
model += x[0] + x[1] + x[2] + x[5] + x[6] >= 22
model += x[0] + x[1] + x[2] + x[3] + x[6] >= 18
model += x[0] + x[1] + x[2] + x[3] + x[4] >= 24
model += x[1] + x[2] + x[3] + x[4] + x[5] >= 12
model += x[2] + x[3] + x[4] + x[5] + x[6] >= 14

# Solve model
model.solve()

#Print model status
print('Status:', LpStatus[model.status])

#Print solution variables
for variable in model.variables():
    print ('{} = {}'.format(variable.name, variable.varValue))

Status: Optimal
workers__0 = 7.0
workers__1 = 7.0
workers__2 = 0.0
workers__3 = 0.0
workers__4 = 10.0
workers__5 = 0.0
workers__6 = 8.0


**Soution: Factory needs to hire 32 workers in total, 7 of them start working on Monday, the other 7 on Tuesday, 10 on Friday, and 8 on Sunday.**